### ISSUE TITLE
Make all fixed-key deserialization methods accept arrays

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bitcoin#3235]
Slices just add one more variant to deal with which is not great. People can convert slices to arrays now and even if you wanted to argue that's nicer for parsing, it's still pretty annoying and error-prone to update the slice after reading so the actual nicer API would be to accept `&mut &[u8]` and update the slice. Guess what that thing also implements. Yes, the `Read` trait. And yes, I do have a first-hand experience with these things. Changing `&[u8]` to `&mut &[u8]` significantly helped with reducing bugs in Firefish.

Even for `PublicKey` I'd love to see an API that just doesn't produce invalid length error for lengths 33 and 65.

**Comment by apoelstra** [comment]
I think since https://github.com/rust-bitcoin/rust-bitcoin/pull/2277 we're ok regarding `PublicKey`. We now have `CompressedPublicKey` for "modern" uses and if `PublicKey` is ugly/annoying to parse then we can live with that.

But concept ACK adding `from_byte_array` methods to everything. Annoyingly we will need to do a long deprecation cycle of the existing methods, which means we'll have to carry the ugly error types (or at least, the now-unused variants) around for a while.

**Comment by apoelstra** [comment]
And right -- I believe right now you can just use `consensus_decode` with a `&mut &[u8]` to get the "update slice when parsing" behavior.

Maybe we should advertise that or provide docs/examples? Up to this point we've sorta treated `consensus_decode` as an obscure part of the library that we don't expect users to need, but personally I use it all the time for reasons like this.

**Comment by Kixunil** [comment]
> Annoyingly we will need to do a long deprecation cycle of the existing methods, which means we'll have to carry the ugly error types (or at least, the now-unused variants) around for a while.

Then I want to have different error types for them. Though it's annoying if we have to give the new error types some silly name. Or maybe put it into the compatibility layer only? Basically `bitcoin0.32.x` would have a feature to depend on keys and impl `Into` on them so one could progressively update them.

**Comment by Kixunil** [comment]
> Maybe we should advertise that or provide docs/examples?

Good idea, judging from how many times I see `Cursor` being used without `Seek` people often miss/forget about `impl Read for &[u8]`. But with `push_decode` we will also have a method that doesn't return the `io::Error` out of the box which is nice too.

**Comment by tcharding** [comment]
> Annoyingly we will need to do a long deprecation cycle of the existing methods

Maybe not :) #3241
